app/lib/task/global_lock.dart

```diff
@@ -0,0 +1,259 @@
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:clock/clock.dart';
+import 'package:logging/logging.dart' show Logger;
+import 'package:pub_dev/shared/datastore.dart';
+import 'package:pub_dev/task/global_lock_models.dart';
+import 'package:ulid/ulid.dart' show Ulid;
+
+final _log = Logger('pub.global_lock');
+
+class GlobalLock {
+  final String _lockId;
+  final Duration _expiration;
+  final DatastoreDB _db;
+
+  GlobalLock._(this._lockId, this._expiration, this._db);
+
+  static GlobalLock create(
+    String lockId, {
+    Duration expiration = const Duration(minutes: 25),
+  }) =>
+      GlobalLock._(lockId, expiration, dbService);
+
+  /// Call [fn] while retaining a claim to this lock. This will wait until the
+  /// lock is acquired.
+  ///
+  /// A [TimeoutException] is thrown if [abort] is completed before the lock is
+  /// acquired.
+  Future<T> withClaim<T>(
+    FutureOr<T> Function(GlobalLockClaim claim) fn, {
+    Completer<void>? abort,
+  }) async {
+    ArgumentError.checkNotNull(fn, 'fn');
+    abort ??= Completer();
+
+    final c = await claim(abort: abort);
+    final claimId = c._entry.claimId;
+    var refreshed = Future.value(true);
+    var done = false;
+    try {
+      scheduleMicrotask(() async {
+        while (c.valid && !done) {
+          // Await for 50% of the time until expiration is gone
+          var delay = c.expires
+              .subtract(_expiration * 0.5)
+              .difference(clock.now().toUtc());
+          // always sleep at-least 10% of expiration
+          if (delay < _expiration * 0.1) {
+            delay = _expiration * 0.1;
+          }
+          await Future.delayed(delay);
+
+          // Try to refresh, if claim is still valid and we're not done.
+          if (c.valid && !done) {
+            refreshed = c.refresh();
+            try {
+              if (!await refreshed) {
+                _log.warning(
+                  'failed to refresh claim $claimId on $_lockId',
+                );
+                return; // Stop refreshing, if refresh failed.
+              }
+            } on Exception catch (e, st) {
+              // Stop refreshing, and log the error
+              _log.severe('Error refreshing claim $claimId on $_lockId', e, st);
+              return;
+            }
+          }
+        }
+      });
+      return await fn(c);
+    } finally {
+      done = true;
+      try {
+        await refreshed; // wait to any ongoing refreshing attempt is done
+      } on Exception {
+        // Ignore the exception, we've logged this before
+      }
+      if (c.valid) {
+        await c.release();
+      }
+    }
+  }
+
+  /// Try to claim the lock, return `null` if lock is currently held by a
+  /// different process.
+  Future<GlobalLockClaim?> tryClaim() async {
```

Comment:
I agree, it would be better to always go through `withClaim`.
But there could be cases where you're looking for a non-blocking lock attempt, and that is what this is for.

Yes, I don't think we'll need it. But it's nice to have exposed for testing, and I see no risk in exposing it. It's pretty much a sane method.